### PR TITLE
fix(volume_server): pin EC shard auto-select to the .ecx-owning disk (#9212)

### LINKED ISSUE #9212: EC volumes with shards distributed across multiple disks on same node are not fully registered with master on startup
**SeaweedFS version:** 4.20 50f25bb5c

**Environment:**
- 4 volume server pods, each with 4 physical disks mounted at `/data1`, `/data2`, `/data3`, `/data4`
- EC encoding: 10+4 (14 shards total)
- Kubernetes deployment

**Description**

When an EC volume has its shards distributed across multiple disks on the same volume server node, the master only registers a subset of those shards upon startup. The remaining shards are physically present on disk but invisible to the master.

At startup, the volume server logs show the following message for affected volumes:

```
EC volume 15: distributed EC (.dat removed) with 2 shards
EC volume 15: distributed EC (.dat removed) with 1 shards
EC volume 15: distributed EC (.dat removed) with 3 shards
```

The message appears multiple times per volume - once per disk that contains shards of that volume. The shard count increments with each disk but the master ultimately only registers a partial count.

**Example - volume 15:**

Physical state on disks:
- `volume-0 /data2`: ec00
- `volume-1 /data1`: ec13
- `volume-1 /data3`: ec03
- `volume-2 /data1`: ec02, ec06, ec10
- `volume-2 /data2`: ec01, ec04, ec09
- `volume-2 /data4`: ec08, ec12
- `volume-3 /data2`: ec05
- `volume-3 /data4`: ec07, ec11

All 14 shards (ec00–ec13) are physically present across the 4 nodes. However, the master reported only 6 shards visible:

```
volume-0: shards [0]
volume-1: shards [3]
volume-2: shards [8, 12]
volume-3: shards [7, 11]
```

This caused the volume to appear as `Critical` in the master topology despite all data being intact on disk.

`ec.rebuild` dry-run confirmed the volume as unrepairable:
```
ec volume 15 is unrepairable with 6 shards (need 10), skipping
```

**Workaround we applied**

We manually moved all shards of each affected volume onto a single disk per node (consolidating to `/data1`), then restarted the volume server pod. After restart, the volume server correctly registered all shards from that single disk location and reported them to the master.

Before consolidation - volume 15 visible to master: **6 shards**
After consolidation and restart - volume 15 visible to master: **11 shards**

The consolidation required also copying the `.vif` file alongside the `.ec??` shards, as without `.vif` the shards were not registered at all.

An additional observation: after restart, SeaweedFS sometimes created an empty `grafana-loki_15.dat` file (0 bytes) on the disk where the `.vif` was placed, which appeared to interfere with EC shard registration on that disk.

**Scale of the problem**

In our cluster with 4 volume nodes each having 4 disks, we found **135 distributed EC volume/node combinations** affected across the `grafana-loki` collection alone. The same pattern was observed in other collections.

The distributed state appears to have been caused by `ec.balance` or `ec.rebuild` operations that moved shards between nodes without accounting for existing shards of the same volume already present on a different disk of the destination node.

**Steps to reproduce**

1. Run `ec.balance` or `ec.rebuild` on a cluster where volume nodes have multiple disks
2. Observe that some EC volumes end up with shards of the same volume on multiple disks of the same node
3. Restart the volume server
4. Query `volume.list -volumeId=X` on master - only a subset of shards is reported despite all being physically present

**Expected behavior**

When a volume server starts and finds shards of the same EC volume across multiple disks, it should register all of them with the master rather than only a partial set.

### PR DESCRIPTION
**Pull request by chrislusf**
## Summary

Follow-up to #9244. Closes the **placement** half of issue [#9212](https://github.com/seaweedfs/seaweedfs/issues/9212): the volume-server auto-select now pins each EC shard to the disk that already owns the volume's `.ecx` index file, even when nothing has been mounted yet.

## Root cause

`weed/shell/command_ec_rebuild.go` only sets `CopyEcxFile=true` for the **first** shard copied to the rebuilder. Subsequent shards arrive with `CopyEcxFile=false` and rely on `VolumeEcShardsCopy` / `ReceiveFile` auto-select to land on the same disk:

```go
location = vs.store.FindFreeLocation(func(loc *storage.DiskLocation) bool {
    _, found := loc.FindEcVolume(needle.VolumeId(req.VolumeId))
    return found
})
```

`FindEcVolume` only consults the **in-memory** `ecVolumes` map. Mid-rebuild nothing has been mounted yet on the destination, so the lookup returns nothing and the fallback picks "any HDD with free space" — which can split shards from their `.ecx` across disks of the same node. That's the orphan-shard layout the reporter sees, fixed on the loader side in #9244.

## What this PR does

- **`Store.FindEcShardTargetLocation(collection, vid)`** — single canonical placement primitive. Selection order:
  1. disk that already has the EC volume mounted,
  2. disk that owns the `.ecx` on disk (volume not yet mounted — the missing case),
  3. any HDD with free space,
  4. any disk with free space.
- **`DiskLocation.HasEcxFileOnDisk(collection, vid)`** — the new on-disk primitive. Checks `IdxDirectory` first with a fallback to `Directory` (covers `.ecx` written before `-dir.idx` was configured).
- **`VolumeEcShardsCopy`** and **`ReceiveFile`** route through `FindEcShardTargetLocation`, dropping their duplicated 4-level fallback ladder.

The existing `needEcxFile` pattern in `command_ec_rebuild.go` (only the first shard carries `.ecx` / `.vif`) is preserved. With placement now pinned to the `.ecx`-owning disk, that optimization remains correct: every shard lands on the disk that received the first shard's `.ecx`, and only one writer ever touches the index files — important because parallel `ec.balance` / `ec.rebuild` operations on the same volume could otherwise race on a shared `.ecx` write (`writeToFile` does `O_TRUNC` then streams).

No protocol changes. Callers that pass an explicit `DiskId` are unaffected.

## Test plan

- [x] `TestFindEcShardTargetLocation_PinsToEcxOnDisk` — drops a sealed `.ecx` on disk 2 (no mount), verifies placement returns disk 2 instead of falling through to disk 0 (the HDD with the most free count). Captures the bug.
- [x] `TestFindEcShardTargetLocation_PrefersMountedOverEcx` — guards that an already-mounted EC volume on disk 1 wins over a stray `.ecx` on disk 2 (steady-state behaviour).
- [x] `TestFindEcShardTargetLocation_FallsThroughToHddWhenNothingMatches` — keeps the existing fallback intact for the cold-volume case.
- [x] `go test ./weed/storage/... ./weed/topology/... ./weed/server/... ./weed/shell/...` all pass; `go vet` clean on touched packages.

## Stacking

Best reviewed after #9244 lands. The two together close #9212 end-to-end:

| PR     | Side                | What it fixes                                                                                       |
|--------|---------------------|-----------------------------------------------------------------------------------------------------|
| #9219  | master              | Aggregates per-disk EC shard messages on the master so all disks of one DataNode are visible.       |
| #9244  | volume-server load  | Loads orphan shards across disks at startup when index files live on a sibling disk.                |
| this   | volume-server place | Stops creating new orphan-shard layouts: pins shard placement to the disk that has `.ecx`.          |

## Summary by CodeRabbit

* **Improvements**
  * Centralized, more consistent erasure-coded shard placement: selection now prefers disks with the EC volume mounted, then disks holding the EC index on disk, then HDDs with capacity, then any disk; free-slot calculations better reflect EC shard capacity.

* **Tests**
  * Added unit tests validating mounted-volume preference, on-disk index preference, HDD fallback, unlimited-capacity behavior, and tight-provisioning edge cases.

### CODE REVIEW

File: weed/storage/store_ec.go [hunk 1]
```diff
@@ -22,6 +22,88 @@ import (
...
+func ecFreeShardCount(loc *DiskLocation, dataShardCount int) int32 {
+	if dataShardCount <= 0 {
+		return 0
+	}
+	free := loc.MaxVolumeCount - int32(loc.VolumesLen())
+	free *= int32(dataShardCount)
+	free -= int32(loc.EcShardCount())
+	free /= int32(dataShardCount)
+	return free
+}
```
[the remark this answers is not in the document]
Comment (chrislusf) [reply weed/storage/store_ec.go:127]: Addressed in e56081cd6. ecFreeShardCount now special-cases MaxVolumeCount<=0 with a synthetic large free count that decrements with current usage, so unlimited disks are eligible and tie-breaks still prefer the less-loaded one. Added TestFindEcShardTargetLocation_HonoursUnlimitedDisk as the regression. Thanks!

File: weed/storage/store_ec.go [hunk 2]
```diff
@@ -22,6 +22,101 @@ import (
...
+			return 1
+		}
+		return unlimitedFree - used
+	}
+	free := loc.MaxVolumeCount - int32(loc.VolumesLen())
+	free *= int32(dataShardCount)
+	free -= int32(loc.EcShardCount())
+	free /= int32(dataShardCount)
+	return free
+}
```
[the remark this answers is not in the document]
Comment (chrislusf) [reply weed/storage/store_ec.go:127]: Addressed in b84322f6a. ecFreeShardCount now returns shard slots directly (drops the trailing /dataShardCount divide), and the unlimited branch is updated to match. Added TestFindEcShardTargetLocation_TightProvisioningKeepsEcxDisk pinning the exact MaxVolumeCount=1 / EcShardCount=1 case from the report. Thanks!

File: weed/storage/store_ec.go [hunk 3]
```diff
@@ -22,6 +22,110 @@ import (
...
+// store (see hasFreeDiskLocation). Reporting a synthetic large free
+// count keeps unlimited disks eligible while still letting tie-breaks
+// prefer the less-loaded one.
+func ecFreeShardCount(loc *DiskLocation, dataShardCount int) int32 {
+	if dataShardCount <= 0 {
+		return 0
+	}
+	if loc.MaxVolumeCount <= 0 {
+		const unlimitedFree = int32(1 << 30)
+		used := int32(loc.VolumesLen())*int32(dataShardCount) + int32(loc.EcShardCount())
```
[the remark this answers is not in the document]
Comment (chrislusf) [reply weed/storage/store_ec.go:115]: `CalculateEcLogicalSize()` does not exist in this repo (`grep` finds no matches), and `EcShardCount()` is the right primitive here: this code computes free *placement capacity* — how many shard slots remain — not logical byte size. Counting shards already on disk is exactly what we want. The intermediate math is bounded by `MaxVolumeCount` (int32) × `dataShardCount` (int) plus `EcShardCount` (returns int), all comfortably within int32 for any realistic disk; no overflow risk. Will hold off on this one — happy to revisit if a real helper lands later.
